File: src/serialized.ts
import { FALSE_NODE, NULL_NODE, TRUE_NODE, type DocNode } from "./document.js";
import { JsonToMarkdownError, type SourceLocation } from "./errors.js";

function locationAt(src: string, offset: number): SourceLocation {
  let line = 1;
  let column = 1;
  const end = Math.min(offset, src.length);
  for (let i = 0; i < end; i++) {
    if (src.charCodeAt(i) === 0x0a) {
      line++;
      column = 1;
    } else {
      column++;
    }
  }
  return { offset, line, column };
}

function syntaxError(src: string, message: string, offset: number): never {
  throw new JsonToMarkdownError("INVALID_JSON_SYNTAX", message, { location: locationAt(src, offset) });
}

function isWs(code: number): boolean {
  return code === 0x20 || code === 0x09 || code === 0x0a || code === 0x0d;
}

function skipWs(src: string, pos: number): number {
  const n = src.length;
  while (pos < n && isWs(src.charCodeAt(pos))) pos++;
  return pos;
}

function isDigit(code: number): boolean {
  return code >= 0x30 && code <= 0x39;
}

interface Scanned {
  readonly node: DocNode;
  readonly end: number;
}

function parseString(src: string, pos: number): { value: string; end: number } {
  const n = src.length;
  let i = pos + 1;
  let start = i;
  let out = "";

  while (i < n) {
    const c = src.charCodeAt(i);
    if (c === 0x22 /* " */) {
      return { value: out + src.slice(start, i), end: i + 1 };
    }
    if (c === 0x5c /* \ */) {
      out += src.slice(start, i);
      i++;
      const e = src[i];
      switch (e) {
        case '"':
          out += '"';
          break;
        case "\\":
          out += "\\";
          break;
        case "/":
          out += "/";
          break;
        case "b":
          out += "\b";
          break;
        case "f":
          out += "\f";
          break;
        case "n":
          out += "\n";
          break;
        case "r":
          out += "\r";
          break;
        case "t":
          out += "\t";
          break;
        case "u": {
          const hex = src.slice(i + 1, i + 5);
          if (hex.length < 4 || !/^[0-9a-fA-F]{4}$/.test(hex)) {
            syntaxError(src, "Invalid \\u escape in string", i - 1);
          }
          out += String.fromCharCode(parseInt(hex, 16));
          i += 4;
          break;
        }
        default:
          syntaxError(src, "Invalid escape sequence in string", i - 1);
      }
      i++;
      start = i;
      continue;
    }
    if (c < 0x20) {
      syntaxError(src, "Unescaped control character in string", i);
    }
    i++;
  }
  return syntaxError(src, "Unterminated string", pos);
}

function parseNumber(src: string, pos: number): Scanned {
  const n = src.length;
  let i = pos;
  if (src.charCodeAt(i) === 0x2d /* - */) i++;

  const first = src.charCodeAt(i);
  if (first === 0x30 /* 0 */) {
    i++;
  } else if (first >= 0x31 && first <= 0x39) {
    i++;
    while (i < n && isDigit(src.charCodeAt(i))) i++;
  } else {
    syntaxError(src, "Invalid number", pos);
  }

  if (i < n && src.charCodeAt(i) === 0x2e /* . */) {
    i++;
    if (!(i < n && isDigit(src.charCodeAt(i)))) syntaxError(src, "Invalid number: missing fraction digits", i);
    while (i < n && isDigit(src.charCodeAt(i))) i++;
  }

  const exp = i < n ? src.charCodeAt(i) : 0;
  if (exp === 0x65 || exp === 0x45 /* e E */) {
    i++;
    const sign = i < n ? src.charCodeAt(i) : 0;
    if (sign === 0x2b || sign === 0x2d) i++;
    if (!(i < n && isDigit(src.charCodeAt(i)))) syntaxError(src, "Invalid number: missing exponent digits", i);
    while (i < n && isDigit(src.charCodeAt(i))) i++;
  }

  return { node: { kind: "number", lexeme: src.slice(pos, i) }, end: i };
}

function parseScalar(src: string, pos: number): Scanned {
  const c = src.charCodeAt(pos);
  if (c === 0x22) {
    const s = parseString(src, pos);
    return { node: { kind: "string", value: s.value }, end: s.end };
  }
  if (c === 0x2d || isDigit(c)) return parseNumber(src, pos);
  if (src.startsWith("true", pos)) return { node: TRUE_NODE, end: pos + 4 };
  if (src.startsWith("false", pos)) return { node: FALSE_NODE, end: pos + 5 };
  if (src.startsWith("null", pos)) return { node: NULL_NODE, end: pos + 4 };
  return syntaxError(src, "Unexpected token; expected a JSON value", pos);
}

type ObjFrame = {
  readonly type: "obj";
  readonly entries: [string, DocNode][];
  readonly seen: Map<string, number>;
  state: "start" | "value" | "comma";
  key: string;
};
type ArrFrame = {
  readonly type: "arr";
  readonly items: DocNode[];
  state: "start" | "value";
};
type Frame = ObjFrame | ArrFrame;

/**
 * Parse exactly one JSON document into a DocNode without executing caller code.
 * Iterative to satisfy the no-nesting-limit ADR; preserves numeric lexemes,
 * member encounter order, and rejects duplicate member names.
 */
export function parseSerialized(source: string): DocNode {
  const n = source.length;
  const stack: Frame[] = [];
  let pending: DocNode | null = null;
  let hasPending = false;
  let pos = skipWs(source, 0);

  if (pos >= n) syntaxError(source, "Unexpected end of input; expected a JSON value", pos);

  // Read the leading value (opens a frame or produces a pending scalar).
  pos = startValue();

  for (;;) {
    const frame = stack[stack.length - 1];

    if (frame === undefined) {
      // Root value complete.
      pos = skipWs(source, pos);
      if (pos !== n) syntaxError(source, "Unexpected trailing content after JSON value", pos);
      return pending!;
    }

    if (frame.type === "arr") {
      if (hasPending) {
        frame.items.push(pending!);
        hasPending = false;
        pending = null;
        frame.state = "value";
        continue;
      }
      pos = skipWs(source, pos);
      const c = source.charCodeAt(pos);
      if (frame.state === "start") {
        if (c === 0x5d /* ] */) {
          pos++;
          closeFrame();
          continue;
        }
        pos = startValue();
        continue;
      }
      // state === "value"
      if (c === 0x2c /* , */) {
        pos = startValue(pos + 1);
        continue;
      }
      if (c === 0x5d) {
        pos++;
        closeFrame();
        continue;
      }
      syntaxError(source, "Expected ',' or ']' in array", pos);
    } else {
      if (hasPending) {
        frame.entries.push([frame.key, pending!]);
        hasPending = false;
        pending = null;
        frame.state = "value";
        continue;
      }
      pos = skipWs(source, pos);
      const c = source.charCodeAt(pos);
      if (frame.state === "start") {
        if (c === 0x7d /* } */) {
          pos++;
          closeFrame();
          continue;
        }
        pos = readMember(frame);
        continue;
      }
      if (frame.state === "value") {
        if (c === 0x2c /* , */) {
          frame.state = "comma";
          pos++;
          continue;
        }
        if (c === 0x7d) {
          pos++;
          closeFrame();
          continue;
        }
        syntaxError(source, "Expected ',' or '}' in object", pos);
      }
      // state === "comma": expect the next member key.
      pos = readMember(frame);
    }
  }

  function startValue(from = pos): number {
    let p = skipWs(source, from);
    if (p >= n) syntaxError(source, "Unexpected end of input; expected a JSON value", p);
    const c = source.charCodeAt(p);
    if (c === 0x7b /* { */) {
      stack.push({ type: "obj", entries: [], seen: new Map(), state: "start", key: "" });
      return p + 1;
    }
    if (c === 0x5b /* [ */) {
      stack.push({ type: "arr", items: [], state: "start" });
      return p + 1;
    }
    const scanned = parseScalar(source, p);
    pending = scanned.node;
    hasPending = true;
    return scanned.end;
  }

  function closeFrame(): void {
    const frame = stack.pop()!;
    pending = frame.type === "arr" ? { kind: "array", items: frame.items } : { kind: "object", entries: frame.entries };
    hasPending = true;
  }

  function readMember(frame: ObjFrame): number {
    let p = skipWs(source, pos);
    if (source.charCodeAt(p) !== 0x22 /* " */) syntaxError(source, "Expected a string object key", p);
    const keyStart = p;
    const key = parseString(source, p);
    const prev = frame.seen.get(key.value);
    if (prev !== undefined) {
      throw new JsonToMarkdownError("DUPLICATE_MEMBER_NAME", `Duplicate object member name ${JSON.stringify(key.value)}.`, {
        location: locationAt(source, keyStart),
        firstLocation: locationAt(source, prev),
      });
    }
    frame.seen.set(key.value, keyStart);
    frame.key = key.value;
    p = skipWs(source, key.end);
    if (source.charCodeAt(p) !== 0x3a /* : */) syntaxError(source, "Expected ':' after object key", p);
    return startValue(p + 1);
  }
}

File: go/serialized.go
package jsontomd

import (
	"bytes"
	"fmt"
	"unicode/utf8"
)

// Port of the TS iterative parser (src/serialized.ts): parses exactly one
// JSON document into a node without recursion, preserving numeric lexemes and
// member encounter order, and rejecting duplicate member names with both
// occurrences located. Positions are byte offsets; they become UTF-16
// SourceLocations lazily, only when an error is built.

func isWsByte(c byte) bool {
	return c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d
}

func skipWs(src []byte, pos int) int {
	for pos < len(src) && isWsByte(src[pos]) {
		pos++
	}
	return pos
}

func isDigit(c byte) bool { return c >= '0' && c <= '9' }

const (
	frameStart uint8 = iota
	frameValue
	frameComma
)

type frame struct {
	isObj   bool
	entries []member
	seen    map[string]int // member name -> key start (byte offset)
	state   uint8
	key     string
	items   []node
}

type parser struct {
	src        []byte
	pos        int
	stack      []frame
	pending    node
	hasPending bool
}

func (p *parser) syntaxErr(message string, offset int) error {
	return &Error{
		Code:     ErrInvalidJSONSyntax,
		Message:  message,
		Location: utf16Location(p.src, offset),
	}
}

// byteAt is charCodeAt with a 0 sentinel past the end: 0 matches none of the
// structural bytes callers test, so out-of-range positions fall through to
// the same syntax errors as the TS NaN comparisons.
func (p *parser) byteAt(pos int) byte {
	if pos < len(p.src) {
		return p.src[pos]
	}
	return 0
}

func isHexDigit(c byte) bool {
	return isDigit(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
}

func hexVal(c byte) rune {
	switch {
	case c <= '9':
		return rune(c - '0')
	case c >= 'a':
		return rune(c-'a') + 10
	default:
		return rune(c-'A') + 10
	}
}

// parseString scans the string starting at the opening quote. The value is
// built as WTF-8: \uXXXX escapes append UTF-16 code units, with adjacent
// high+low escape pairs combining into one supplementary rune exactly as the
// units would join in a JS string, and lone surrogates surviving.
func (p *parser) parseString(pos int) (value string, end int, err error) {
	src := p.src
	n := len(src)
	i := pos + 1
	start := i
	var out []byte
	pendingHigh := rune(-1)

	flushPending := func() {
		if pendingHigh >= 0 {
			out = appendWTF8(out, pendingHigh)
			pendingHigh = -1
		}
	}

	for i < n {
		c := src[i]
		if c == '"' {
			flushPending()
			if out == nil {
				return string(src[start:i]), i + 1, nil
			}
			return string(append(out, src[start:i]...)), i + 1, nil
		}
		if c == '\\' {
			// A raw run and a pending high cannot coexist (any raw byte after
			// an escape flushes below), so append order here is safe.
			out = append(out, src[start:i]...)
			i++
			var e byte
			if i < n {
				e = src[i]
			}
			if e == 'u' {
				if i+4 >= n || !isHexDigit(src[i+1]) || !isHexDigit(src[i+2]) || !isHexDigit(src[i+3]) || !isHexDigit(src[i+4]) {
					return "", 0, p.syntaxErr(`invalid \u escape in string`, i-1)
				}
				unit := hexVal(src[i+1])<<12 | hexVal(src[i+2])<<8 | hexVal(src[i+3])<<4 | hexVal(src[i+4])
				switch {
				case unit >= 0xdc00 && unit <= 0xdfff:
					if pendingHigh >= 0 {
						// Adjacent high+low escapes join into one rune, as the
						// UTF-16 units would join in a JS string.
						out = utf8.AppendRune(out, (pendingHigh-0xd800)<<10|(unit-0xdc00)+0x10000)
						pendingHigh = -1
					} else {
						out = appendWTF8(out, unit)
					}
				case unit >= 0xd800 && unit <= 0xdbff:
					flushPending()
					pendingHigh = unit
				default:
					flushPending()
					out = utf8.AppendRune(out, unit)
				}
				i += 4
			} else {
				flushPending()
				switch e {
				case '"', '\\', '/':
					out = append(out, e)
				case 'b':
					out = append(out, '\b')
				case 'f':
					out = append(out, '\f')
				case 'n':
					out = append(out, '\n')
				case 'r':
					out = append(out, '\r')
				case 't':
					out = append(out, '\t')
				default:
					return "", 0, p.syntaxErr("invalid escape sequence in string", i-1)
				}
			}
			i++
			start = i
			// A pending high survives only into an immediately following
			// escape; a raw byte or the closing quote fixes it as lone.
			if pendingHigh >= 0 && (i >= n || src[i] != '\\') {
				flushPending()
			}
			continue
		}
		if c < 0x20 {
			return "", 0, p.syntaxErr("unescaped control character in string", i)
		}
		i++
	}
	return "", 0, p.syntaxErr("unterminated string", pos)
}

func (p *parser) parseNumber(pos int) (node, int, error) {
	src := p.src
	n := len(src)
	i := pos
	if p.byteAt(i) == '-' {
		i++
	}

	first := p.byteAt(i)
	if first == '0' {
		i++
	} else if first >= '1' && first <= '9' {
		i++
		for i < n && isDigit(src[i]) {
			i++
		}
	} else {
		return node{}, 0, p.syntaxErr("invalid number", pos)
	}

	if i < n && src[i] == '.' {
		i++
		if !(i < n && isDigit(src[i])) {
			return node{}, 0, p.syntaxErr("invalid number: missing fraction digits", i)
		}
		for i < n && isDigit(src[i]) {
			i++
		}
	}

	if e := p.byteAt(i); e == 'e' || e == 'E' {
		i++
		if s := p.byteAt(i); s == '+' || s == '-' {
			i++
		}
		if !(i < n && isDigit(src[i])) {
			return node{}, 0, p.syntaxErr("invalid number: missing exponent digits", i)
		}
		for i < n && isDigit(src[i]) {
			i++
		}
	}

	return node{kind: kindNumber, lexeme: string(src[pos:i])}, i, nil
}

func (p *parser) parseScalar(pos int) (node, int, error) {
	c := p.byteAt(pos)
	if c == '"' {
		value, end, err := p.parseString(pos)
		if err != nil {
			return node{}, 0, err
		}
		return node{kind: kindString, str: value}, end, nil
	}
	if c == '-' || isDigit(c) {
		return p.parseNumber(pos)
	}
	rest := p.src[pos:]
	if bytes.HasPrefix(rest, []byte("true")) {
		return trueNode, pos + 4, nil
	}
	if bytes.HasPrefix(rest, []byte("false")) {
		return falseNode, pos + 5, nil
	}
	if bytes.HasPrefix(rest, []byte("null")) {
		return nullNode, pos + 4, nil
	}
	return node{}, 0, p.syntaxErr("unexpected token; expected a JSON value", pos)
}

// startValue reads the next value beginning at or after `from`: it opens a
// container frame or produces a pending scalar, leaving p.pos after it.
func (p *parser) startValue(from int) error {
	pos := skipWs(p.src, from)
	if pos >= len(p.src) {
		return p.syntaxErr("unexpected end of input; expected a JSON value", pos)
	}
	switch p.src[pos] {
	case '{':
		p.stack = append(p.stack, frame{isObj: true, seen: make(map[string]int)})
		p.pos = pos + 1
		return nil
	case '[':
		p.stack = append(p.stack, frame{})
		p.pos = pos + 1
		return nil
	}
	scanned, end, err := p.parseScalar(pos)
	if err != nil {
		return err
	}
	p.pending = scanned
	p.hasPending = true
	p.pos = end
	return nil
}

func (p *parser) closeFrame() {
	f := p.stack[len(p.stack)-1]
	p.stack = p.stack[:len(p.stack)-1]
	if f.isObj {
		p.pending = node{kind: kindObject, entries: f.entries}
	} else {
		p.pending = node{kind: kindArray, items: f.items}
	}
	p.hasPending = true
}

func (p *parser) readMember(f *frame) error {
	pos := skipWs(p.src, p.pos)
	if p.byteAt(pos) != '"' {
		return p.syntaxErr("expected a string object key", pos)
	}
	keyStart := pos
	key, end, err := p.parseString(pos)
	if err != nil {
		return err
	}
	if prev, dup := f.seen[key]; dup {
		return &Error{
			Code:          ErrDuplicateMemberName,
			Message:       fmt.Sprintf("duplicate object member name %q", key),
			Location:      utf16Location(p.src, keyStart),
			FirstLocation: utf16Location(p.src, prev),
		}
	}
	f.seen[key] = keyStart
	f.key = key
	pos = skipWs(p.src, end)
	if p.byteAt(pos) != ':' {
		return p.syntaxErr("expected ':' after object key", pos)
	}
	return p.startValue(pos + 1)
}

// parseSerialized parses exactly one JSON document without executing caller
// code, per the semantics shared with the TS implementation.
func parseSerialized(src []byte) (node, error) {
	p := &parser{src: src}
	p.pos = skipWs(src, 0)
	if p.pos >= len(src) {
		return node{}, p.syntaxErr("unexpected end of input; expected a JSON value", p.pos)
	}
	if err := p.startValue(p.pos); err != nil {
		return node{}, err
	}

	for {
		if len(p.stack) == 0 {
			p.pos = skipWs(src, p.pos)
			if p.pos != len(src) {
				return node{}, p.syntaxErr("unexpected trailing content after JSON value", p.pos)
			}
			return p.pending, nil
		}
		f := &p.stack[len(p.stack)-1]

		if !f.isObj {
			if p.hasPending {
				f.items = append(f.items, p.pending)
				p.hasPending = false
				p.pending = node{}
				f.state = frameValue
				continue
			}
			p.pos = skipWs(src, p.pos)
			c := p.byteAt(p.pos)
			if f.state == frameStart {
				if c == ']' {
					p.pos++
					p.closeFrame()
					continue
				}
				if err := p.startValue(p.pos); err != nil {
					return node{}, err
				}
				continue
			}
			if c == ',' {
				if err := p.startValue(p.pos + 1); err != nil {
					return node{}, err
				}
				continue
			}
			if c == ']' {
				p.pos++
				p.closeFrame()
				continue
			}
			return node{}, p.syntaxErr("expected ',' or ']' in array", p.pos)
		}

		if p.hasPending {
			f.entries = append(f.entries, member{name: f.key, value: p.pending})
			p.hasPending = false
			p.pending = node{}
			f.state = frameValue
			continue
		}
		p.pos = skipWs(src, p.pos)
		c := p.byteAt(p.pos)
		if f.state == frameStart {
			if c == '}' {
				p.pos++
				p.closeFrame()
				continue
			}
			if err := p.readMember(f); err != nil {
				return node{}, err
			}
			continue
		}
		if f.state == frameValue {
			if c == ',' {
				f.state = frameComma
				p.pos++
				continue
			}
			if c == '}' {
				p.pos++
				p.closeFrame()
				continue
			}
			return node{}, p.syntaxErr("expected ',' or '}' in object", p.pos)
		}
		// frameComma: expect the next member key.
		if err := p.readMember(f); err != nil {
			return node{}, err
		}
	}
}
